cpp/test/sg/rf_treelite_test.cu

```diff
@@ -0,0 +1,416 @@
+    float missing_value = FLT_MAX;
+    TREELITE_CHECK(TreeliteAssembleDenseBatch(
+      inference_data_h.data(), missing_value, params.n_inference_rows,
+      params.n_cols, &dense_batch));
+
+    // Use dense batch so batch_sparse is 0.
+    // pred_margin = true means to produce raw margins rather than transformed probability.
+    int batch_sparse = 0;
+    bool pred_margin = false;
+    // Allocate larger array for treelite predicted label with using multi-class classification to aviod seg faults.
+    // Altough later we only use first params.n_inference_rows elements.
+    size_t treelite_predicted_labels_size;
+
+    TREELITE_CHECK(TreelitePredictorPredictBatch(
+      predictor, dense_batch, batch_sparse, verbose, pred_margin,
+      treelite_predicted_labels.data(), &treelite_predicted_labels_size));
+
+    TREELITE_CHECK(TreeliteDeleteDenseBatch(dense_batch));
+    TREELITE_CHECK(TreelitePredictorFree(predictor));
+  }
+
+  void epsilonCheck() {
+    float epsilon = std::numeric_limits<float>::epsilon();
+
+    diff_elements = 0;
+    // To convert the probablily of binary classification to class index
+    // The number in treelite_label is the probability of selecting class 1 instead of class 0
+    // So we could say if the treelite_label is larger than 0.5, then it predicts class 1
+    // There is a problem when probability equals to 0.5 (there is a tie between two classes)
+    // Here return class 1 in this case but in Rapids RF, we use the class which first reachs the half votes as the final class:
+    // https://gitlab-master.nvidia.com/RAPIDS/cuml/blob/branch-0.9/cpp/src/randomforest/randomforest_impl.cuh#L291
+    for (int i = 0; i < params.n_inference_rows; i++) {
+      if (is_classification) {
+        treelite_predicted_labels[i] =
+          treelite_predicted_labels[i] >= 0.5 ? 1 : 0;
+      }
+
+      if (ref_predicted_labels[i] - treelite_predicted_labels[i] > epsilon) {
+        diff_elements++;
+      }
+    }
+  }
+
+  void SetUp() override { commonSetUp(); }
+
+  void TearDown() override {
+    CUDA_CHECK(cudaStreamDestroy(stream));
+
+    data_h.clear();
+    inference_data_h.clear();
+    treelite_predicted_labels.clear();
+    ref_predicted_labels.clear();
+
+    CUDA_CHECK(cudaFree(data_d));
+    CUDA_CHECK(cudaFree(inference_data_d));
+  }
+
+ protected:
+  RfInputs<T> params;
+  RF_params rf_params;
+
+  T *data_d, *inference_data_d;
+  std::vector<T> data_h;
+  std::vector<T> inference_data_h;
+
+  int diff_elements = 0;
+  // Set to 1 for regression and 2 for binary classification
+  // #class for multi-classification
+  int task_category;
+  int is_classification;
+
+  int data_len;
+  int inference_data_len;
+
+  cudaStream_t stream;
+  cumlHandle handle;
+
+  std::vector<float> treelite_predicted_labels;
+  std::vector<float> ref_predicted_labels;
+};
+
+template <typename T>
+class RfTreeliteTestClf : public RfTreeliteTestCommon<T> {
+ protected:
+  void testClassifier() {
```

Comment:
`testClassifier()` and `testRegressor()` contain a lot of common code. Consider moving it to the base class, and leaving only the parts that differ in the derived classes.

Reply:
Fixed, now the `testClassifier()` and `testRegressor()` only have code to set values and update that to GPU

Reply:
There is still some duplication remaining. Not all of it can be eliminated, populating data and inference data could be moved to the base class.